server/sql_info_fetcher.go

```diff
@@ -0,0 +1,308 @@
+)
+
+type sqlInfoFetcher struct {
+	store tikv.Storage
+	do    *domain.Domain
+	s     session.Session
+}
+
+type tableNamePair struct {
+	DBName    string
+	TableName string
+}
+
+type tableNameExtractor struct {
+	curDB string
+	names map[tableNamePair]struct{}
+}
+
+func (tne *tableNameExtractor) Enter(in ast.Node) (ast.Node, bool) {
+	if _, ok := in.(*ast.TableName); ok {
+		return in, true
+	}
+	return in, false
+}
+
+func (tne *tableNameExtractor) Leave(in ast.Node) (ast.Node, bool) {
+	if t, ok := in.(*ast.TableName); ok {
+		tp := tableNamePair{DBName: t.Schema.L, TableName: t.Name.L}
+		if tp.DBName == "" {
+			tp.DBName = tne.curDB
+		}
+		if _, ok := tne.names[tp]; !ok {
+			tne.names[tp] = struct{}{}
+		}
+	}
+	return in, true
+}
+
+func (sh *sqlInfoFetcher) zipInfoForSQL(w http.ResponseWriter, r *http.Request) {
+	var err error
+	sh.s, err = session.CreateSession(sh.store)
+	if err != nil {
+		serveError(w, http.StatusInternalServerError, fmt.Sprintf("create session failed, err: %v", err))
+		return
+	}
+	defer sh.s.Close()
+	sh.do = domain.GetDomain(sh.s)
+	reqCtx := r.Context()
+	sql := r.FormValue("sql")
+	pprofTimeString := r.FormValue("pprof_time")
+	timeoutString := r.FormValue("timeout")
+	curDB := strings.ToLower(r.FormValue("current_db"))
+	if curDB != "" {
+		_, err = sh.s.Execute(reqCtx, fmt.Sprintf("use %v", curDB))
+		if err != nil {
+			serveError(w, http.StatusInternalServerError, fmt.Sprintf("use database failed, err: %v", err))
+			return
+		}
+	}
+	var (
+		pprofTime int
+		timeout   int
+	)
+	if pprofTimeString != "" {
+		pprofTime, err = strconv.Atoi(pprofTimeString)
+	}
+	if err != nil {
+		serveError(w, http.StatusBadRequest, "invalid value for pprof_time")
+		return
+	}
+	if pprofTimeString != "" && pprofTime < 5 {
+		serveError(w, http.StatusBadRequest, "pprof time is too short")
+	}
+	if timeoutString != "" {
+		timeout, err = strconv.Atoi(timeoutString)
+	}
+	if err != nil {
+		serveError(w, http.StatusBadRequest, "invalid value for timeout")
+		return
+	}
+	if timeout < pprofTime {
+		timeout = pprofTime
+	}
+	pairs, err := sh.extractTableNames(sql, curDB)
+	if err != nil {
+		serveError(w, http.StatusBadRequest, fmt.Sprintf("invalid SQL text, err: %v", err))
+		return
+	}
+	zw := zip.NewWriter(w)
+	defer func() {
+		terror.Log(zw.Close())
+	}()
+	for pair := range pairs {
+		jsonTbl, err := sh.getStatsForTable(pair)
+		if err != nil {
+			terror.Log(err)
+			continue
+		}
+		statsFw, err := zw.Create(fmt.Sprintf("%v.%v.json", pair.DBName, pair.TableName))
+		if err != nil {
+			terror.Log(err)
+			continue
+		}
+		data, err := json.Marshal(jsonTbl)
+		if err != nil {
+			terror.Log(err)
+			continue
+		}
+		_, err = statsFw.Write(data)
+		if err != nil {
+			terror.Log(err)
+			continue
+		}
+	}
+	for pair := range pairs {
+		err = sh.getShowCreateTable(pair, zw)
+		if err != nil {
+			serveError(w, http.StatusInternalServerError, fmt.Sprintf("get schema for %v.%v failed", pair.DBName, pair.TableName))
```

Comment:
Is it ok that we write error messages and zip formatted data to `w` at the same time?

Reply:
In fact, it's not very ok.
```
➜  tidb git:(http-fetch-sql-info) ✗ curl -X POST http://127.0.0.1:10080/debug/sub-optimal-plan -d "sql=select%20*%20from%20t&current_db=test"
execute SQL failed, err: [planner:1046]No database selected
PK
test.t.json��Aj�0���Z
                     ��i��� [iDe+X������"=C��a��[h�Ϋ?���� V4#��陉�6����Y��Ƣ�}��2
u�B��P�:��Ω��t�⧤���~���ܣ&�Ѭ��ɕ�A�E�t�wc�y�k�m�zl6��aw|�yC����H�؆"v�F�+����ן�7?j\���u�O��P� "f���Ӡa\����T
XQtest.t.schema.txt*�tru
                        qqt�qUH(IP��RPH�LP(K,J�H,�02�Tpqus
                                                          �	Q�
                                                                  ����Tp�s��s�����wq�K:{8��ؖ��Y�&�(8���8�����I�y\���P(%��ip� "f
XQ
  test.t.json(%��ipCtest.t.schema.txtPKx�%
```
But the api `debug/zip` did the same thing. https://github.com/pingcap/tidb/pull/9651
Or we may need to write to a BytesBuffer first then write to ResponseWriter

Reply:
OK... for simplifying, we can leave it as a further optimization and do it in another PR.